include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -672,27 +672,35 @@ template <typename _Event, typename... _Args>
 class __future : private std::tuple<_Args...>
 {
     _Event __my_event;
+    bool __waited_for_result = false;
 
     template <typename _T>
     constexpr auto
     __wait_and_get_value(const sycl::buffer<_T>& __buf)
     {
         //according to a contract, returned value is one-element sycl::buffer
-        return __buf.get_host_access(sycl::read_only)[0];
+        auto __val = __buf.get_host_access(sycl::read_only)[0];
+        __set_waited_for_result();
+        return __val;
     }
 
     template <typename _ExecutionPolicy, typename _T>
     constexpr auto
     __wait_and_get_value(const __result_and_scratch_storage<_ExecutionPolicy, _T>& __storage)
     {
-        return __storage.__wait_and_get_value(__my_event);
+        auto __val = __storage.__wait_and_get_value(__my_event);
+        __set_waited_for_result();
+        return __val;
     }
 
     template <typename _T>
     constexpr auto
     __wait_and_get_value(const _T& __val)
     {
         wait();
```

Comment:
Do we need to wait here again even if the user has called `__future.wait()`? Could we instead check for `__waited_for_result` and skip the wait if it is true.

Reply:
My assumption that `__waited_for_result` state is debugging tool only, may be we remove it at all later or place under some `debug` mode macro check. So I believe that we shouldn't use it for choice how to get result and so on. With this tool I am trying to find errors in tests now on `CPU` and on `GPU`.

Reply:
as result, all `waited_for_result` stuff has been removed to reduce the amount of changes.

Reply:
@julianmi may we resolve this conversation?